### ISSUE TITLE
[RyuJIT/arm32][JitStressRegs=4] Assertion failed 'isDoubleReg(reg1)'

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#14200]
Seen in a Windows x86_arm_altjit test run with `COMPlus_JitStressRegs=4`.

Tests:
```
JIT\Regression\JitBlue\GitHub_13056\GitHub_13056\GitHub_13056.cmd
JIT\Performance\CodeQuality\BenchmarksGame\n-body\n-body-3\n-body-3.cmd
```

Example:
```
Running: E:\gh\coreclr2\bin\tests\Windows_NT.x86.Release\JIT\Regression\JitBlue\GitHub_13056\GitHub_13056\GitHub_13056.cmd -coreroot E:\gh\coreclr2\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root
BEGIN EXECUTION
 "E:\gh\coreclr2\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\corerun.exe" GitHub_13056.exe

Assert failure(PID 20804 [0x00005144], Thread: 11832 [0x2e38]): Assertion failed 'isDoubleReg(reg1)' in 'Regression13056:Clamp(double,double,double):double' (IL size 25)

    File: e:\gh\coreclr2\src\jit\emitarm.cpp Line: 2096
    Image: E:\gh\coreclr2\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\CoreRun.exe

Expected: 100
Actual: 123456789
END EXECUTION - FAILED
FAILED
```

**Comment by BruceForstall** [comment]
cc @dotnet/arm32-contrib

**Comment by YongseopKim** [comment]
With the latest commit(8c7049aabe4bd2d5332a795bfa61a37077ba39bc) + PRI2

```
 dragon@dragon-rasp3  ~/dotnet/arm32-dotnet  uname -a
Linux dragon-rasp3 4.1.19-v7+ #858 SMP Tue Mar 15 15:56:00 GMT 2016 armv7l armv7l armv7l GNU/Linux
```

```
 dragon@dragon-rasp3  ~/dotnet/arm32-dotnet  cat tests.txt 
JIT/Regression/JitBlue/GitHub_13056/GitHub_13056
JIT/Performance/CodeQuality/BenchmarksGame/n-body/n-body-3
```

```
 ✘ dragon@dragon-rasp3  ~/dotnet/arm32-dotnet  ./run_clrutc_ryujit_arm.sh
LOG: CLR_RYUJIT_10161526.log
export COMPlus_AltJit="*"
export COMPlus_JitStressRegs=4
/home/dragon/dotnet/arm32-dotnet/coreclr/tests/runtest.sh -v --show-time --coreOverlayDir=/home/dragon/dotnet/arm32-dotnet/tmp/coreoverlay-ARM-CHE-REL-10161107/overlay --testRootDir=/home/dragon/dotnet/tcs/ARM_Rel_CLEARED_TC_Pri2 --testNativeBinDir=/home/dragon/dotnet/arm32-dotnet/tmp/coreoverlay-ARM-CHE-REL-10161107/overlay --testDirFile=./tests.txt
Running on  CPU- arm
Skipping crossgen of FX assemblies.
Skip preparing for GC stress test. Dependent package is not supported on this architecture.
Test directory does not exist: #Exceptions/ForeignThread/ForeignThreadExceptions
Test directory does not exist: #GC/Scenarios/ServerModel/servermodel
Test directory does not exist: #JIT/jit64/opt/cse/hugeexpr1
Test directory does not exist: #JIT/jit64/opt/cse/HugeField1
Test directory does not exist: #JIT/Methodical/doublearray/dblarray3_cs_d
Test directory does not exist: #JIT/Methodical/doublearray/dblarray3_cs_do
Test directory does not exist: #JIT/Methodical/doublearray/dblarray3_cs_r
Test directory does not exist: #JIT/Methodical/doublearray/dblarray3_cs_ro
Test directory does not exist: #
Preparing JIT/Regression/JitBlue/GitHub_13056/GitHub_13056/GitHub_13056.sh
The tests have been prepared
Starting JIT/Regression/JitBlue/GitHub_13056/GitHub_13056/GitHub_13056.sh
Test directory does not exist: JIT/Performance/CodeQuality/BenchmarksGame/n-body/n-body-3
FAILED   - [   0][   5s]JIT/Regression/JitBlue/GitHub_13056/GitHub_13056/GitHub_13056.sh
               BEGIN EXECUTION
               /home/dragon/dotnet/arm32-dotnet/tmp/coreoverlay-ARM-CHE-REL-10161107/overlay/corerun GitHub_13056.exe
               
               Assert failure(PID 12979 [0x000032b3], Thread: 12979 [0x32b3]): Assertion failed 'isDoubleReg(reg1)' in 'Regression13056:Clamp(double,double,double):double' (IL size 25)
               
                   File: /home/dragon/dotnet/arm32-dotnet/coreclr/src/jit/emitarm.cpp Line: 2096
                   Image: /home/dragon/dotnet/arm32-dotnet/tmp/coreoverlay-ARM-CHE-REL-10161107/overlay/corerun
               
               ./GitHub_13056.sh: line 243: 12979 Aborted                 (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
               Expected: 100
               Actual: 134
               END EXECUTION - FAILED

=======================
     Test Results
=======================
# CoreCLR Bin Dir  : 
# Tests Discovered : 1
# Passed           : 0
# Failed           : 1
# Skipped          : 0
=======================
0 minutes and 26 seconds taken to run CoreCLR tests.
```

Now only GitHub_13056 is failed in linux_arm32_ryujit.

**Comment by YongseopKim** [comment]
CodeGen
```
Added IP mapping: 0x000F STACK_EMPTY (G_M9195_IG03,ins#3,ofs#14) label
Generating: N027 (  2,  2) [000029] -------------                IL_OFFSET void   IL offset: 0xf REG NA
Generating: N029 (  1,  1) [000027] -------------       t27 =    LCL_VAR   double V01 arg1         u:2 f30 (last use) REG f30 $81
                                                              ┌──▌  t27    double
Generating:                [000045] -------------       t45 = ▌  COPY      double REG f1         // HERE
                                                              ┌──▌  t45    double
Generating: N031 (  2,  2) [000028] -------------             ▌  RETURN    double REG NA $144
                            V01 in reg f30 is becoming dead  [000027]
                            Live regs: 0000 {f30 f31} => 0000 {}
                            Live vars: {V01} => {}
```

The regnum `f1` causes the assertion.

```c++
// emitarm.cpp
void emitter::emitIns_R_R(
    instruction ins, emitAttr attr, regNumber reg1, regNumber reg2, insFlags flags /* = INS_FLAGS_DONT_CARE */)

{
    emitAttr  size = EA_SIZE(attr);
    insFormat fmt  = IF_NONE;
    insFlags  sf   = INS_FLAGS_DONT_CARE;

    /* Figure out the encoding format of the instruction */
    switch (ins)
    {
        // ...
        case INS_vabs:
        case INS_vsqrt:
        case INS_vcmp:
        case INS_vneg:
            if (size == EA_8BYTE)
            {
                assert(isDoubleReg(reg1));  // HERE
                assert(isDoubleReg(reg2));
            }
            else
            {
                assert(isFloatReg(reg1));
                assert(isFloatReg(reg2));
            }
            __fallthrough;

// emitarm.h
inline static bool isDoubleReg(regNumber reg)
{
    return isFloatReg(reg) && ((reg % 2) == 0);
}
```

**Comment by YongseopKim** [comment]
In LSRA,
```
*************** In LinearScan::allocateRegisters()

Linear scan intervals before allocateRegisters:
Interval  0: (V00) RefPositions {#2@0 #7@21 #12@41 #17@59} physReg:f0 Preferences=[f0]
Interval  1: (V01) RefPositions {#0@0 #4@9 #8@21 #10@31 #20@67} physReg:f2 Preferences=[f0]
Interval  2: (V02) RefPositions {#1@0 #5@9 #13@41 #15@51 #24@71} physReg:f4 Preferences=[f0-f1]
Interval  3: RefPositions {#22@68 #31@77} physReg:NA Preferences=[f0]
Interval  4: RefPositions {#26@72 #33@77} physReg:NA Preferences=[f2]
Interval  5: (constant) RefPositions {#27@74 #28@75} physReg:NA Preferences=[allInt]
Interval  6: RefPositions {#29@76 #34@77} physReg:NA Preferences=[allInt]

------------
REFPOSITIONS BEFORE ALLOCATION:
------------
...
<RefPosition #10  @31  ->#20  RefTypeUse <Ivl:1 V01> LCL_VAR BB03 regmask=[f0-f1] last>
...
VAR REFPOSITIONS BEFORE ALLOCATION
...
--- V01
<RefPosition #0   @0   ->#4   RefTypeParamDef <Ivl:1 V01> BB00 regmask=[f2] fixed>
<RefPosition #4   @9   ->#8   RefTypeUse <Ivl:1 V01> LCL_VAR BB01 regmask=[allDouble]>
<RefPosition #8   @21  ->#10  RefTypeUse <Ivl:1 V01> LCL_VAR BB02 regmask=[allDouble]>
<RefPosition #10  @31  ->#20  RefTypeUse <Ivl:1 V01> LCL_VAR BB03 regmask=[f0-f1] last>
<RefPosition #20  @67  RefTypeUse <Ivl:1 V01> LCL_VAR BB07 regmask=[f0] last fixed>
...


Allocating Registers
──────────────────────────────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┤
LocRP# Name Type  Action Reg  │r0  │r1  │r2  │r3  │r4  │r5  │r6  │r7  │r8  │r9  │r10 │r11 │r12 │sp  │lr  │pc  │f0  │f1  │f2  │f3  │f4  │f5  │f6  │f7  │f8  │f9  │f10 │f11 │f12 │f13 │f14 │f15 │f16 │f17 │f18 │f19 │f20 │f21 │f22 │f23 │f24 │f25 │f26 │f27 │f28 │f29 │f30 │
──────────────────────────────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┤
25.#9  BB3 PredBB2            │    │    │    │    │----│----│----│----│----│----│----│----│    │----│    │----│V0 i│V0 i│    │    │    │    │    │    │    │    │    │    │    │    │    │    │----│----│----│----│----│----│----│----│----│----│----│----│V2 i│V2 i│V1 a│
31.#10 V1   Use *  Copy  f1   │    │    │    │    │----│----│----│----│----│----│----│----│    │----│    │----│V1 a│V1 a│    │    │    │    │    │    │    │    │    │    │    │    │    │    │----│----│----│----│----│----│----│----│----│----│----│----│V2 i│V2 i│V1 a│
──────────────────────────────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┤


------------
REFPOSITIONS AFTER ALLOCATION:
------------
...
<RefPosition #10  @31  ->#20  RefTypeUse <Ivl:1 V01> LCL_VAR BB03 regmask=[f1] last copy>
...

VAR REFPOSITIONS AFTER ALLOCATION
--- V01
<RefPosition #0   @0   ->#4   RefTypeParamDef <Ivl:1 V01> BB00 regmask=[f30] fixed>
<RefPosition #4   @9   ->#8   RefTypeUse <Ivl:1 V01> LCL_VAR BB01 regmask=[f30]>
<RefPosition #8   @21  ->#10  RefTypeUse <Ivl:1 V01> LCL_VAR BB02 regmask=[f30]>
<RefPosition #10  @31  ->#20  RefTypeUse <Ivl:1 V01> LCL_VAR BB03 regmask=[f1] last copy>
<RefPosition #20  @67  RefTypeUse <Ivl:1 V01> LCL_VAR BB07 regmask=[f0] last copy fixed>

Trees after linear scan register allocator (LSRA)

------------ BB03 [00F..011) (return), preds={BB02} succs={}
N027 (  2,  2) [000029] -------------                IL_OFFSET void   IL offset: 0xf REG NA
N029 (  1,  1) [000027] -------------       t27 =    LCL_VAR   double V01 arg1         u:2 f30 (last use) REG f30 $81
                                                  ┌──▌  t27    double
               [000045] -------------       t45 = ▌  COPY      double REG f1
                                                  ┌──▌  t45    double
N031 (  2,  2) [000028] -------------             ▌  RETURN    double REG NA $144

Final allocation
──────────────────────────────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┤
LocRP# Name Type  Action Reg  │r0  │r1  │r2  │r3  │r4  │r5  │r6  │r7  │r8  │r9  │r10 │r11 │r12 │sp  │lr  │pc  │f0  │f1  │f2  │f3  │f4  │f5  │f6  │f7  │f8  │f9  │f10 │f11 │f12 │f13 │f14 │f15 │f16 │f17 │f18 │f19 │f20 │f21 │f22 │f23 │f24 │f25 │f26 │f27 │f28 │f29 │f30 │
──────────────────────────────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┤
25.#9  BB3 PredBB2            │    │    │    │    │----│----│----│----│----│----│    │----│    │----│    │----│    │    │    │    │    │    │    │    │    │    │    │    │    │    │    │    │----│----│----│----│----│----│----│----│----│----│----│----│    │----│V1 a│
31.#10 V1   Use *  Copy  f1   │    │    │    │    │----│----│----│----│----│----│    │----│    │----│    │----│    │V1 i│    │    │    │    │    │    │    │    │    │    │    │    │    │    │----│----│----│----│----│----│----│----│----│----│----│----│    │----│V1 i│
──────────────────────────────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┼────┤

BB03 [00F..011) (return), preds={BB02} succs={}
=====
Predecessor for variable locations: BB02
Var=Reg beg of BB03: V01=f30
  N027.                    IL_OFFSET IL offset: 0xf REG NA
  N029.                    V01(f30*)
  N000. f1*             =  COPY     ; f30*
  N031.                    RETURN   ; f1*
Var=Reg end of BB03: none
```

**Comment by YongseopKim** [comment]
The process of the reg alloc for `COPY` has something wrong.

**Comment by sjsinju** [comment]
@YongseopKim I think it can be fixed https://github.com/dotnet/coreclr/pull/14536

**Comment by YongseopKim** [comment]
@sjsinju Yes, I've seen the above tcs are passed with #14536 . @BruceForstall ,  this issue can be closed after merging #14536 .

**Comment by CarolEidt** [comment]
Fixed with #14573
